src/Catalyst.Core.Modules.Keystore/KeyChain.cs

```diff
@@ -0,0 +1,497 @@
+        /// <remarks>
+        ///   The key id is the SHA-256 multihash of its public key. The public key is 
+        ///   a protobuf encoding containing a type and 
+        ///   the DER encoding of the PKCS SubjectPublicKeyInfo.
+        /// </remarks>
+        MultiHash CreateKeyId(AsymmetricKeyParameter key)
+        {
+            var spki = SubjectPublicKeyInfoFactory
+               .CreateSubjectPublicKeyInfo(key)
+               .GetDerEncoded();
+
+            // Add protobuf cruft.
+            var publicKey = new Lib.Cryptography.Proto.PublicKey
+            {
+                Data = spki
+            };
+            if (key is RsaKeyParameters)
+                publicKey.Type = Lib.Cryptography.Proto.KeyType.RSA;
+            else if (key is ECPublicKeyParameters)
+                publicKey.Type = Lib.Cryptography.Proto.KeyType.Secp256k1;
+            else if (key is Ed25519PublicKeyParameters)
+                publicKey.Type = Lib.Cryptography.Proto.KeyType.Ed25519;
+            else
+                throw new NotSupportedException($"The key type {key.GetType().Name} is not supported.");
+
+            using (var ms = new MemoryStream())
+            {
+                ProtoBuf.Serializer.Serialize(ms, publicKey);
+
+                // If the length of the serialized bytes <= 42, then we compute the "identity" multihash of 
+                // the serialized bytes. The idea here is that if the serialized byte array 
+                // is short enough, we can fit it in a multihash verbatim without having to 
+                // condense it using a hash function.
+                var alg = (ms.Length <= 48) ? "identity" : "sha2-256";
+
+                ms.Position = 0;
+                return MultiHash.ComputeHash(ms, alg);
+            }
+        }
+
+        AsymmetricCipherKeyPair GetKeyPairFromPrivateKey(AsymmetricKeyParameter privateKey)
+        {
+            AsymmetricCipherKeyPair keyPair = null;
+            if (privateKey is RsaPrivateCrtKeyParameters rsa)
+            {
+                var pub = new RsaKeyParameters(false, rsa.Modulus, rsa.PublicExponent);
+                keyPair = new AsymmetricCipherKeyPair(pub, privateKey);
+            }
+            else if (privateKey is Ed25519PrivateKeyParameters ed)
+            {
+                var pub = ed.GeneratePublicKey();
+                keyPair = new AsymmetricCipherKeyPair(pub, privateKey);
+            }
+            else if (privateKey is ECPrivateKeyParameters ec)
+            {
+                var q = ec.Parameters.G.Multiply(ec.D);
+                var pub = new ECPublicKeyParameters(ec.AlgorithmName, q, ec.PublicKeyParamSet);
+                keyPair = new AsymmetricCipherKeyPair(pub, ec);
+            }
+
+            if (keyPair == null)
+                throw new NotSupportedException($"The key type {privateKey.GetType().Name} is not supported.");
+
+            return keyPair;
+        }
+
+        class PasswordFinder : IPasswordFinder, IDisposable
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)